core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,402 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.Flags;
+
+import io.netty.channel.EventLoop;
+
+/**
+ * A {@link StreamMessage} optimized for when writes and reads all happen on the provided {@link EventLoop},
+ * removing the need for most synchronization from the hot path. This should satisfy standard cases when using
+ * Armeria. It is not required for writes or reads to use the provided {@link EventLoop} but in such a case, it
+ * will be significantly faster to use {@link DefaultStreamMessage}.
+ *
+ * <p>Note that when {@link Subscription#cancel()} or {@link #abort()} are called from a different thread, the
+ * stream will continue to signal objects until demand is satisfied, rather than stopping in the middle. If this
+ * is an issue, it is recommended to use {@link DefaultStreamMessage}.
+ *
+ * <p>Methods in this class prefixed with 'do' must be run on the stream's event loop.
+ */
+public class EventLoopStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<EventLoopStreamMessage> abortedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(EventLoopStreamMessage.class, "aborted");
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<EventLoopStreamMessage> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(EventLoopStreamMessage.class, "subscribed");
+
+    private static final Logger logger = LoggerFactory.getLogger(EventLoopStreamMessage.class);
+
+    private final EventLoop eventLoop;
+    private final Queue<Object> queue;
+
+    private SubscriptionImpl subscription;
+    private long demand;
+    private boolean invokedOnSubscribe;
+    private boolean inOnNext;
+
+    private State state = State.OPEN;
+
+    @SuppressWarnings("unused")
+    private volatile int subscribed;  // set only via subscribedUpdater
+    @SuppressWarnings("unused")
+    private volatile int aborted;  // set only via abortedUpdater
+    private volatile boolean isOpen = true;
+    private volatile boolean wroteAny;
+
+    /**
+     * Creates a new {@link EventLoopStreamMessage} which executes all writes on an arbitrary {@link EventLoop}.
+     * It is highly recommended to use {@link #EventLoopStreamMessage(EventLoop)} instead to allow writes to
+     * happen on the same {@link EventLoop} as this stream's.
+     */
+    public EventLoopStreamMessage() {
+        this(CommonPools.workerGroup().next());
+        logger.debug("Creating EventLoopStreamMessage without specifying EventLoop. This will be very slow.");
+    }
+
+    /**
+     * Creates a new {@link EventLoopStreamMessage} which executes all writes on the provided {@link EventLoop}.
+     */
+    public EventLoopStreamMessage(EventLoop eventLoop) {
+        this.eventLoop = requireNonNull(eventLoop, "eventLoop");
```

Comment:
Could we determine the event loop rather at subscription time? e.g.

```java
public void subscribe(Subscriber ...) { // no executor specified
    EventLoop eventLoop = RequestContext.mapCurrent(
            RequestContext::eventLoop, () -> CommonPools.workerGroup().next());
}
```

Reply:
What if we use an MPSC queue from jctools? Not as fast as `ArrayDequeue` but may be good enough for low-contention?

Reply:
Read your comment after writing this comment. Now I understand why it's a good idea to specify an `EventLoop` at construction time. See [my comment](https://github.com/line/armeria/pull/855#issuecomment-346567012).